Add Tdk invensense icp201xx pressure sensor support

rbuisson-invn opened the pull request:
drivers: sensor: icp201xx: supports icp201xx sensors
dts: supports icp201xx invn pressure sensor
samples: sensor: pressure_polling: barometric pressure, temperature and altitude generic sample
samples: sensor: pressure_interrupt: barometric pressure sensor interrupts generic sample
module: HAL TDK module

Adds official TDK Invensense Inc. driver in TDK HAL module for icp201xx sensor.
Add generic samples supporting this driver.
Validated with custom setup: nrf52dk_nrf52832 + icp201xx EVB board
Build ok by running:
west twister -T samples\sensor\pressure_polling -T samples\sensor\pressure_interrupt -p nrf52dk/nrf52832

Signed-off-by: Remi Buisson [remi.buisson@tdk.com](mailto:remi.buisson@tdk.com)

rbuisson-invn commented:
Hi @MaureenHelm ,
This PR is waiting a review for 3 weeks.
It is blocking further developments at our side.
I know it's quite a big chunk of code, but we would be very grateful if you could take a look at it.
Regards,

yperess commented on dts/bindings/sensor/invensense,icp201xx.yaml:
```diff
@@ -0,0 +1,46 @@
+    default: mode0
+    description: |
+      Default mode for sensor.
+      Operation mode as described in the documentation.
```
What document? Please provide a link here or add the details in the description

yperess commented on drivers/sensor/tdk/icp201xx/icp201xx_trigger.c:
```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright (c) 2020 TDK Invensense
```
License years are old

rbuisson-invn replied:
Updated, thanks

yperess commented on drivers/sensor/tdk/icp201xx/icp201xx_spi.c:
```diff
@@ -0,0 +1,56 @@
+
+#include "icp201xx_drv.h"
+
+#if CONFIG_SPI
```
You already check this in your CMakeLists.txt, no need to add the condition here

yperess commented on drivers/sensor/tdk/icp201xx/icp201xx_i2c.c:
```diff
@@ -0,0 +1,31 @@
+
+#include "icp201xx_drv.h"
+
+#if CONFIG_I2C
```
Same here, this condition is already checked in the CMakeLists.txt

rbuisson-invn replied:
Correct, removed!

yperess commented on drivers/sensor/tdk/icp201xx/icp201xx_drv.h:
```diff
@@ -0,0 +1,92 @@
+	uint8_t drive_strength;
+};
+
+#ifdef CONFIG_ICP201XX_TRIGGER
```
You can leave these in the header regardless of the CONFIG, makes the header more readable.

rbuisson-invn replied:
Sure, done.

yperess commented on drivers/sensor/tdk/icp201xx/icp201xx_drv.h:
```diff
@@ -0,0 +1,92 @@
+	icp201xx_reg_write_fn write;
+};
+
+#if CONFIG_SPI
```
Same here, don't wrap with `#if`, if there are no references to the variable it'll just get stripped.

yperess replied:
Ping

yperess commented on drivers/sensor/tdk/icp201xx/icp201xx_drv.c:
```diff
@@ -0,0 +1,449 @@
+
+#define IO_DRIVE_STRENGTH_ADDR 0x0D
+
+void inv_icp201xx_sleep_us(int us)
```
Maybe add `static inline` here?

yperess commented on drivers/sensor/tdk/icp201xx/icp201xx_drv.c:
```diff
@@ -0,0 +1,449 @@
+
+LOG_MODULE_REGISTER(ICP201XX, CONFIG_SENSOR_LOG_LEVEL);
+
+#ifdef CONFIG_CPU_HAS_FPU
+#include <math.h>
+#define ATMOSPHERICAL_PRESSURE_KPA ((float)101.325)
+#define TO_KELVIN(temp_C)          (((float)273.15) + temp_C)
+/* Constant in altitude formula:
+ * M*g/R = (0,0289644 * 9,80665 / 8,31432)
+ * with M the molar mass of air.
+ * with g the gravitational force acceleration.
+ * with R the universal gaz constant.
+ */
+#define HEIGHT_TO_PRESSURE_COEFF   ((float)0.03424)
+
+/* Constant in altitude formula:
+ * R / (M*g) = 8,31432 / (0,0289644 * 9,80665)
+ * with M the molar mass of air.
+ * with g the gravitational force acceleration.
+ * with R the universal gaz constant.
+ */
+#define PRESSURE_TO_HEIGHT_COEFF   ((float)29.27127)
+/* Constant for altitude formula:
+ * logarithm of pressure at 0m
+ * ln(101.325)
+ */
+#define LOG_ATMOSPHERICAL_PRESSURE ((float)4.61833)
+#endif
```
Please consider using the `dsp.h` `q31_t` here

rbuisson-invn replied:
At some point, it requires pure floats to compute logarithm.
I don't think dsp.h provides such support.

yperess replied:
Does it? The DSP header provide mechanism for multiplying fixed points which can be just as accurate. I'm OK with you using floats, but it does limit this driver to `CONFIG_FPU`.

yperess replied:
I still don't think you actually need a float, you should be able to use a fixed point API.

yperess commented on drivers/sensor/tdk/icp201xx/icp201xx_drv.c:
```diff
@@ -0,0 +1,449 @@
+	inv_icp201xx_app_warmup(&(data->icp_device), data->op_mode, ICP201XX_MEAS_MODE_CONTINUOUS);
+
+#ifdef CONFIG_ICP201XX_TRIGGER
+	if (config->gpio_int.port) {
```
You're already checking for `config->gpio_int.port` inside the `trigget_init` function. This can be simplified to:

```c
if (IS_ENABLED(CONFIG_ICP201XX_TRIGGER)) {
  rc = icp201xx_trigger_init(dev);
  if (rc < 0) {
    LOG_ERR("Failed to initialize interrupt.");
    return rc;
  }
}
```

rbuisson-invn replied:
Correct, fixed.

yperess commented on drivers/sensor/tdk/icp201xx/icp201xx_drv.c:
```diff
@@ -0,0 +1,449 @@
+		&icp201xx_api_funcs);
+
+#define DT_DRV_COMPAT invensense_icp201xx
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
```
No need for this `#if`, `DT_INST_FOREACH_STATUS_OKAY` will do nothing if there are no `status="okay"` nodes.

rbuisson-invn replied:
Ok, removed.

yperess commented on drivers/sensor/tdk/icp201xx/Kconfig:
```diff
@@ -0,0 +1,53 @@
+		Enable driver for ICP101xx Barometric Pressure and Temperature Sensor.
+
+if ICP201XX
+config ICP201XX_TRIGGER
+	bool
+	default y
+	depends on GPIO
+	depends on $(dt_compat_any_has_prop,$(DT_COMPAT_INVENSENSE_ICP201XX),int-gpios)
+
+choice ICP201XX_TRIGGER_MODE
+	prompt "Trigger mode"
+	default ICP201XX_TRIGGER_GLOBAL_THREAD
+	depends on ICP201XX_TRIGGER
+	help
+		Specify the type of triggering to be used by the driver.
+
+config ICP201XX_TRIGGER_GLOBAL_THREAD
+	bool "Use global thread"
+
+config ICP201XX_TRIGGER_OWN_THREAD
+	bool "Use own thread"
+
+endchoice
+
+config ICP201XX_THREAD_PRIORITY
+	int "Thread priority"
+	depends on ICP201XX_TRIGGER_OWN_THREAD
+	default 10
+	help
+	  Priority of thread used by the driver to handle interrupts.
+
+config ICP201XX_THREAD_STACK_SIZE
+	int "Thread stack size"
+	depends on ICP201XX_TRIGGER_OWN_THREAD
+	default 1024
+	help
+		Stack size of thread used by the driver to handle interrupts.
```
Should use https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/Kconfig.trigger_template instead of manually defining these.

rbuisson-invn replied:
Thanks for the hints, done !

yperess commented on drivers/sensor/tdk/icp201xx/Kconfig:
```diff
@@ -0,0 +1,53 @@
+	help
+		Enable driver for ICP101xx Barometric Pressure and Temperature Sensor.
+
+if ICP201XX
```
Consider adding logging via https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/logging/Kconfig.template.log_config

rbuisson-invn replied:
The driver, only reports LOG_ERROR, I don't think there is a real benefit in using a specific module LOG management.
I may be wrong on this, so don't hesitate to correct me if my understanding is bad :)

rbuisson-invn commented:
Thanks for your review @yperess !

@MaureenHelm and @teburd this PR is for the second TDK pressure sensor (having some extra features compared to https://github.com/zephyrproject-rtos/zephyr/pull/84081 you already reviewed).
Would you mind taking a quick look at this one too?
Thanks :)

kartben commented on samples/sensor/pressure_polling/sample.yaml:
```diff
@@ -0,0 +1,9 @@
+  sample.sensor.pressure_sensor:
+    build_only: true
+    tags: sensors
+    filter: dt_alias_exists("pressure_sensor")
```
This currently doesn't run. Needs the following change, but also keep only of the two overlay and name it `nrf52dk_nrf52832.overlay` (otherwise neither gets picked up, hence no pressure-sensor alias is found, hence test is filtered :) )

```suggestion
    filter: dt_alias_exists("pressure-sensor")
```

kartben replied:
> This currently doesn't run. Needs the following change, **but also keep only of the two overlays and name it nrf52dk_nrf52832.overlay (otherwise neither gets picked up, hence no pressure-sensor alias is found, hence test is filtered :) )**

not addressed

kartben commented on samples/sensor/pressure_interrupt/sample.yaml:
```diff
@@ -0,0 +1,9 @@
+  sample.sensor.pressure_interrupt:
+    build_only: true
+    tags: sensors
+    filter: dt_alias_exists("pressure_sensor")
```
ditto :)

rbuisson-invn replied:
Thanks, fixed.

kartben reviewed (changes requested):
please also add relevant entries to build_all/sensors dtsi's and conf files as this is the only way to be certain we'll have coverage independently of the existence of samples and respective overlays. 

Thanks!

kartben commented on samples/sensor/pressure_interrupt/README.rst:
```diff
@@ -0,0 +1,72 @@
+.. zephyr:code-sample: pressure_interrupt
```
doesn't show up in samples listing due to typo

Also, varisou formatting issues with the file, please check https://builds.zephyrproject.io/zephyr/pr/83914/docs/samples/sensor/pressure_interrupt/README.html -- applies to the other README too :)

```suggestion
.. :zephyr:code-sample: pressure_interrupt
```

rbuisson-invn replied:
Missing colon is probably not the suggested one :), but it was actually missing!
Thanks, fixed.

kartben commented on samples/sensor/pressure_interrupt/README.rst:
```diff
@@ -0,0 +1,72 @@
+Overview
+********
+
+This sample application uses a pressure sensor interrupt line to:
+** Inform when a measure is available in the sensor FIFO.
+Temperature and pressure data are read and displayed in terminal.
+If floats are supported, estimated altitude is also displayed.
+** Inform when the pressure value crosses a specified threshold.
+Threshold corresponds to around a 50cm altitude increase.
+A message is displayed in the terminal.
+** Inform when the pressure value changed more than the specified
+value between two consecutive samples.
+Change value corresponds to a finger pressing the sensor.
+A message is displayed in the terminal.
```
```suggestion
This sample application uses a pressure sensor interrupt line to:

* Inform when a measure is available in the sensor FIFO.
Temperature and pressure data are read and displayed in terminal.
If floats are supported, estimated altitude is also displayed.
* Inform when the pressure value crosses a specified threshold.
Threshold corresponds to around a 50cm altitude increase.
A message is displayed in the terminal.
* Inform when the pressure value changed more than the specified
value between two consecutive samples.
Change value corresponds to a finger pressing the sensor.
A message is displayed in the terminal.
```

kartben commented on samples/sensor/pressure_interrupt/README.rst:
```diff
@@ -0,0 +1,72 @@
+
+.. code-block:: console
+
+## Default configuration
+
+[00:00:00.266,479] <inf> PRESS_INT_SAMPLE: Starting ICP201xx sample.
+[00:00:00.273,803] <inf> PRESS_INT_SAMPLE: temp 25.49 Cel, pressure 96.271438 kPa, altitude 447.208465 m
+[00:00:00.280,914] <inf> PRESS_INT_SAMPLE: temp 25.50 Cel, pressure 96.271331 kPa, altitude 447.234161 m
+[00:00:00.288,024] <inf> PRESS_INT_SAMPLE: temp 25.49 Cel, pressure 96.266685 kPa, altitude 447.636077 m
+[00:00:00.295,135] <inf> PRESS_INT_SAMPLE: temp 25.50 Cel, pressure 96.267951 kPa, altitude 447.537078 m
+[00:00:00.302,246] <inf> PRESS_INT_SAMPLE: temp 25.51 Cel, pressure 96.268577 kPa, altitude 447.488281 m
+[00:00:00.309,356] <inf> PRESS_INT_SAMPLE: temp 25.50 Cel, pressure 96.269340 kPa, altitude 447.414978 m
+[00:00:00.316,467] <inf> PRESS_INT_SAMPLE: temp 25.50 Cel, pressure 96.268562 kPa, altitude 447.473663 m
+[00:00:00.323,547] <inf> PRESS_INT_SAMPLE: temp 25.50 Cel, pressure 96.267341 kPa, altitude 447.596496 m
+<repeats endlessly>
+
+<when sensor is pressed>
+[00:00:09.819,061] <inf> PRESS_INT_SAMPLE: PRESSURE CHANGE INTERRUPT
+
+<when the sensor pressure crosses defined threhold>
+[00:00:09.859,039] <inf> PRESS_INT_SAMPLE: PRESSURE THRESHOLD INTERRUPT
```
needs 3-space indenting

rbuisson-invn replied:
Thanks, fixed.

kartben commented on samples/sensor/pressure_interrupt/README.rst:
```diff
@@ -0,0 +1,72 @@
+
+.. code-block:: devicetree
+
+ / {
+	aliases {
+			pressure-sensor = &icp201xx;
+		};
+	};
```
```suggestion
	/ {
		aliases {
			pressure-sensor = &icp201xx;
		};
	};
```

rbuisson-invn replied:
Thanks, fixed.

yperess commented on drivers/sensor/tdk/icp201xx/icp201xx_drv.c:
```diff
@@ -0,0 +1,445 @@
+	return cfg->bus_io->write(&cfg->bus, reg, (uint8_t *)wbuffer, wlen);
+}
+
+#ifdef CONFIG_CPU_HAS_FPU
```
Just because the CPU has an FPU doesn't mean that the user wants it enabled, use `CONFIG_FPU` instead.

rbuisson-invn replied:
Done, thanks!

yperess commented on drivers/sensor/tdk/icp201xx/icp201xx_drv.c:
```diff
@@ -0,0 +1,445 @@
+		icp201xx_convert_pressure(val, data->raw_pressure);
+	} else if (chan == SENSOR_CHAN_AMBIENT_TEMP) {
+		icp201xx_convert_temperature(val, data->raw_temperature);
+#ifdef CONFIG_CPU_HAS_FPU
```
Same here, use `CONFIG_FPU`

rbuisson-invn replied:
Thanks for the ping I missed that change.
Thanks, fixed.

rbuisson-invn commented:
@yperess, it's not clear to me how I could compute a logarithm only using fixed points.
The C logf() function is called there: 
drivers/sensor/tdk/icp201xx/icp201xx_drv.c line 75.
static float convertToHeight(float pressure_kp, float temperature_C)
{
	return PRESSURE_TO_HEIGHT_COEFF * TO_KELVIN(temperature_C) *
	       (LOG_ATMOSPHERICAL_PRESSURE - **logf(pressure_kp)**);
}

yperess commented:
> @yperess, it's not clear to me how I could compute a logarithm only using fixed points. The C logf() function is called there: drivers/sensor/tdk/icp201xx/icp201xx_drv.c line 75. static float convertToHeight(float pressure_kp, float temperature_C) { return PRESSURE_TO_HEIGHT_COEFF * TO_KELVIN(temperature_C) * (LOG_ATMOSPHERICAL_PRESSURE - **logf(pressure_kp)**); }

I see we haven't ported this part of the math functions to the DSP subsystem yet. I'll get on that for the next release. The API will look like https://github.com/zephyrproject-rtos/cmsis-dsp/blob/main/Include/dsp/fast_math_functions.h#L158.

Please address the issue in https://github.com/zephyrproject-rtos/zephyr/pull/83914#issuecomment-2666005682 and I'll be happy to approve as is.

kartben commented on tests/drivers/build_all/sensor/app.overlay:
```diff
@@ -147,7 +147,8 @@
 				   <&test_gpio 0 0>,
 				   <&test_gpio 0 0>,
-				   <&test_gpio 0 0>;	/* 0x30 */
+				   <&test_gpio 0 0>;
+				   <&test_gpio 0 0>;	/* 0x31 */
```
I think the original intent here is to give a visual clue every 10 lines, so probably needs to be:

```suggestion
				   <&test_gpio 0 0>;	/* 0x30 */
				   <&test_gpio 0 0>;
```

rbuisson-invn replied:
> I think the original intent here is to give a visual clue every 10 lines, so probably needs to be:

Thanks for active follow up ;)
I'll fix.

kartben reviewed (commented):
looks great, thank you! Just a final nit

rbuisson-invn commented:
> > @yperess, it's not clear to me how I could compute a logarithm only using fixed points. The C logf() function is called there: drivers/sensor/tdk/icp201xx/icp201xx_drv.c line 75. static float convertToHeight(float pressure_kp, float temperature_C) { return PRESSURE_TO_HEIGHT_COEFF * TO_KELVIN(temperature_C) * (LOG_ATMOSPHERICAL_PRESSURE - **logf(pressure_kp)**); }
> 
> I see we haven't ported this part of the math functions to the DSP subsystem yet. I'll get on that for the next release. The API will look like https://github.com/zephyrproject-rtos/cmsis-dsp/blob/main/Include/dsp/fast_math_functions.h#L158.
> 
> Please address the issue in [#83914 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/83914#issuecomment-2666005682) and I'll be happy to approve as is.

Sure, it's now done.
Thanks for the overall review and feedback!

rbuisson-invn commented:
@kartben or @yperess , could you please merge this PR?
Looks like I cannot do it myself.

kartben commented:
> @kartben or @yperess , could you please merge this PR?
> Looks like I cannot do it myself.

It'll have to wait until Zephyr 4.2 merge window opens after March 7 since we're currently in feature freeze for 4.1.

rbuisson-invn commented:
> > @kartben or @yperess , could you please merge this PR?
> > Looks like I cannot do it myself.
> 
> It'll have to wait until Zephyr 4.2 merge window opens after March 7 since we're currently in feature freeze for 4.1.

Ok thanks !

rbuisson-invn commented:
> phyr 4.2 merge window opens after March 7 since we

@kartben or @yperess , could you please merge this PR?
Zephyr release 4.1 seems to be done, but merging is still blocked on my side.

MaureenHelm commented on drivers/sensor/tdk/icp201xx/icp201xx_drv.c:
```diff
@@ -0,0 +1,445 @@
+#endif
+}
+
+static const struct sensor_driver_api icp201xx_api_funcs = {.sample_fetch = icp201xx_sample_fetch,
```
Need to use `DEVICE_API`

rbuisson-invn replied:
Thanks, fixed!

MaureenHelm commented on drivers/sensor/tdk/icp201xx/icp201xx_drv.c:
```diff
@@ -0,0 +1,445 @@
+		COND_CODE_1(DT_INST_ON_BUS(inst, spi),                               \
+		(ICP201XX_CONFIG_SPI(inst)),	                                       \
+		(ICP201XX_CONFIG_I2C(inst)));                                        \
+	DEVICE_DT_INST_DEFINE(inst, icp201xx_init, NULL, &icp201xx_drv_##inst, \
```
Need to use `SENSOR_DEVICE_DT_INST_DEFINE`

rbuisson-invn replied:
Thanks, fixed!

MaureenHelm commented on drivers/sensor/tdk/icp201xx/icp201xx_trigger.c:
```diff
@@ -0,0 +1,219 @@
+	if (i_status) {
+		if ((i_status & ICP201XX_INT_STATUS_FIFO_WMK_HIGH) &&
+		    (drv_data->drdy_handler != NULL)) {
+			static struct sensor_trigger trigger = {.chan = SENSOR_CHAN_ALL,
+								.type = SENSOR_TRIG_DATA_READY};
+			drv_data->drdy_handler(dev, &trigger);
+		}
+		if ((i_status & ICP201XX_INT_STATUS_PRESS_DELTA) &&
+		    (drv_data->delta_handler != NULL)) {
+			static struct sensor_trigger trigger = {.chan = SENSOR_CHAN_PRESS,
+								.type = SENSOR_TRIG_DELTA};
+			drv_data->delta_handler(dev, &trigger);
+		}
+		if ((i_status & ICP201XX_INT_STATUS_PRESS_ABS) &&
+		    (drv_data->threshold_handler != NULL)) {
+			static struct sensor_trigger trigger = {.chan = SENSOR_CHAN_PRESS,
+								.type = SENSOR_TRIG_THRESHOLD};
```
Need to store a pointer to the user-provided trigger and return it here. See #55130

rbuisson-invn replied:
Thanks, fixed!

MaureenHelm commented on drivers/sensor/tdk/icp201xx/icp201xx_trigger.c:
```diff
@@ -0,0 +1,219 @@
+#if defined(CONFIG_ICP201XX_TRIGGER_OWN_THREAD)
+	k_sem_init(&drv_data->gpio_sem, 0, K_SEM_MAX_LIMIT);
+
+	k_thread_create(&drv_data->thread, drv_data->thread_stack,
+			CONFIG_ICP201XX_THREAD_STACK_SIZE, icp201xx_thread, (struct device *)dev,
+			NULL, NULL, K_PRIO_COOP(CONFIG_ICP201XX_THREAD_PRIORITY), 0, K_NO_WAIT);
```
Suggest giving the thread a name with `k_thread_name_set`

rbuisson-invn replied:
Thanks, fixed!

MaureenHelm commented on dts/bindings/sensor/invensense,icp201xx.yaml:
```diff
@@ -0,0 +1,51 @@
+
+  op-mode:
+    type: string
+    default: mode0
```
Need to explain in the description why the default value was selected
https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#dt-bindings-default-rules

MaureenHelm commented on dts/bindings/sensor/invensense,icp201xx.yaml:
```diff
@@ -0,0 +1,51 @@
+
+  drive-strength:
+    type: string
+    default: current_12mA_1_8V
```
Need to explain in the description why the default value was selected
https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#dt-bindings-default-rules

rbuisson-invn replied:
Thanks, fixed!

MaureenHelm commented on tests/drivers/build_all/sensor/sensors_trigger_global.conf:
```diff
@@ -67,3 +67,4 @@ CONFIG_VCNL36825T_TRIGGER_GLOBAL_THREAD=y
 CONFIG_VCNL4040_TRIGGER_GLOBAL_THREAD=y
 CONFIG_WSEN_PADS_2511020213301_TRIGGER_GLOBAL_THREAD=y
 CONFIG_WSEN_TIDS_2521020222501_TRIGGER_GLOBAL_THREAD=y
+CONFIG_ICP201XX_TRIGGER_GLOBAL_THREAD=y
```
Move up to preserve alpha order

rbuisson-invn replied:
Thanks, fixed!

MaureenHelm commented on tests/drivers/build_all/sensor/sensors_trigger_none.conf:
```diff
@@ -68,3 +68,4 @@ CONFIG_VCNL36825T_TRIGGER_NONE=y
 CONFIG_VCNL4040_TRIGGER_NONE=y
 CONFIG_WSEN_PADS_2511020213301_TRIGGER_NONE=y
 CONFIG_WSEN_TIDS_2521020222501_TRIGGER_NONE=y
+CONFIG_ICP201XX_TRIGGER_NONE=y
```
Move up to preserve alpha order

rbuisson-invn replied:
Thanks, fixed!

MaureenHelm commented on tests/drivers/build_all/sensor/sensors_trigger_own.conf:
```diff
@@ -64,3 +64,4 @@ CONFIG_VCNL36825T_TRIGGER_OWN_THREAD=y
 CONFIG_VCNL4040_TRIGGER_OWN_THREAD=y
 CONFIG_WSEN_PADS_2511020213301_TRIGGER_OWN_THREAD=y
 CONFIG_WSEN_TIDS_2521020222501_TRIGGER_OWN_THREAD=y
+CONFIG_ICP201XX_TRIGGER_OWN_THREAD=y
```
Move up to preserve alpha order

rbuisson-invn replied:
Fixed

yperess commented on drivers/sensor/tdk/icp201xx/icp201xx_drv.c:
```diff
@@ -0,0 +1,445 @@
+#ifdef CONFIG_ICP201XX_TRIGGER
+		} else if (attr == SENSOR_ATTR_SLOPE_TH) {
+			data->pressure_change = *val;
+		} else if ((attr == SENSOR_ATTR_LOWER_THRESH) ||
+			   (attr == SENSOR_ATTR_UPPER_THRESH)) {
+			if (val->val1 > 0) {
+				data->pressure_threshold = *val;
```
Why is `pressure_threshold` both a lower and upper threshold?

rbuisson-invn replied:
Because there is only a single threshold that generates the interrupt, so I considered that LOWER and UPPER were equal.
Do you see any better way to handle that feature?

yperess replied:
Does the sensor fire the interrupt any time you go above or below the threshold? Keep in mind that when we're reviewing the code we don't have the device available so we rely on you to tell us how it works. If you set a threshold and get an interrupt every time the pressure goes above and every time you go below then this looks good. So it's the directionality that matters

rbuisson-invn replied:
Thanks, and sorry for the lack of implementation details in my previous question.
Yes, the interrupt is triggered when the pressure value goes above and below the threshold.

rbuisson-invn replied:
If you agree with this implementation (considering that crossing the threshold value in any direction generates an interrupt) could you please re-approve the PR?
(or let me know if anything should be fixed)

ubieda replied:
@yperess Thanks for clarifying - I agree that's what I had understood the sensor does.

What I'm wondering is, though -  if an user will get confused by this behavior:
1. User sets SENSOR_ATTR_LOWER_THRESH for a given threshold.
2. Threshold is crossed by the value going over the threshold (not under).
3. User gets a trigger callback for the given configuration.

yperess replied:
Yeah, it's confusing for sure. I can't come up with a much better solution. I was toying with the idea of setting a software threshold, basically save the lower/upper threshold numbers in the `data` struct then after reading the interrupt decide if we want to actually notify the user. The issue there is that if the user sets both the upper and lower thresholds, the driver can't use the threshold property, it would have to use a delta and get a lot of interrupts potentially.

With that, I think we can agree that you can either set the lower OR upper threshold on this driver. I'm ok with this if that's what you'd like. We can keep the threshold type around and in the interrupt handler (after reading the sample) check
1. if we set an upper threshold and the reading is higher, notify the callback
2. if we set a lower threshold and the reading is lower, notify the callback
3. ignore the interrupt in all other cases

rbuisson-invn replied:
Thanks for this improvement proposal!
The issue here might be that, as the ICP201xx parts do not feature any "direct data register", checking the current pressure will "consumes" data from the FIFO.
If the FIFO interrupt is enabled as well, it will miss some data when Threshold interrupt is served.
I would suggest that the decision to "consume" data, to decide whether it is a lower or upper threshold interrupt, is left to the application.
Please don't hesitate to ask for more clarification if needed.

rbuisson-invn commented:
@MaureenHelm could you please check the changes I've done per your request, and approve if everything is ok for you?

ubieda commented on drivers/sensor/tdk/icp201xx/icp201xx_drv.c:
```diff
@@ -0,0 +1,445 @@
+	icp201xx_mutex_lock(dev);
+
+	/** Read measurements count in FIFO **/
+	if (inv_icp201xx_get_fifo_count(&(data->icp_device), &fifo_packets)) {
```
Nit: For Fetch/Get, Instead of servicing the FIFO: Why not just get the latest readings?

To consider for Sensor Model v2: If this were Read/Decode, structuring a single-shot reading like this would break an on-going data-stream.

yperess replied:
Good catch, no need to read the FIFO if you're going to throw away all but the latest.

rbuisson-invn commented:
@yperess @ubieda, I tried to answer your concerns. Is there still anything gating your approval?
Thanks !

rbuisson-invn commented:
@yperess, your review status is still on "request changes", could please you review my answers and approve if they make sense to you?

rbuisson-invn committed:
modules: hal_tdk: Add icp201xx support

icp201xx are barometric pressure and temperature sensors.
https://invensense.tdk.com/smartpressure

Signed-off-by: Remi Buisson <remi.buisson@tdk.com>

rbuisson-invn committed:
drivers: sensor: icp201xx: Add icp201xx support

Use official TDK Invensense driver for icp201xx sensor in tdk_hal module.

Signed-off-by: Remi Buisson <remi.buisson@tdk.com>

rbuisson-invn committed:
dts: bindings: sensor: Add invensense icp201xx properties

Add I2C and SPI variants for icp201xx.

Signed-off-by: Remi Buisson <remi.buisson@tdk.com>

rbuisson-invn committed:
samples: sensor: Barometric pressure polling generic example

Polls and reports pressure, temperature and altitude data

Signed-off-by: Remi Buisson <remi.buisson@tdk.com>

rbuisson-invn committed:
samples: sensor: Barometric pressure interrupt generic example

Reports generic barometric sensor interrupts
* Data ready interrupt
* Pressure threshold interrupt
* Pressure changed interrupt

Signed-off-by: Remi Buisson <remi.buisson@tdk.com>

rbuisson-invn commented:
@yperess  @teburd  @ubieda @MaureenHelm : This PR required a rebase after your validation, sorry to ask you again :(

rbuisson-invn commented:
@kartben Thanks ;)

kartben commented:
@rbuisson-invn can you please take care of https://github.com/zephyrproject-rtos/zephyr/actions/runs/14310267206/job/40103313166#step:11:1

kartben commented:
> @rbuisson-invn can you please take care of [zephyrproject-rtos/zephyr/actions/runs/14310267206/job/40103313166#step:11:1](https://github.com/zephyrproject-rtos/zephyr/actions/runs/14310267206/job/40103313166#step:11:1)

nevermind - https://github.com/zephyrproject-rtos/zephyr/pull/88266
